### PR TITLE
Opencart 4 Support

### PR DESCRIPTION
**Pull request by nikhilsharma1100**
### - Link added should be a valid URL and accessible throughout the org.
   ### - If the branch name contains hotfix / revert by default the BVT workflow check will pass.

| Test Case Document URL                        |
|-----------------------------------------------|
| Please paste test case document link here.... |

### CODE REVIEW

File: admin/view/template/payment/razorpay.twig [hunk 1]
```diff
@@ -0,0 +1,150 @@
+{{ header }}{{ column_left }}
+<div id="content">
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply admin/view/template/payment/razorpay.twig:2]: @sofiya02 This view file is for admin save settings page. In opencart 3 I don't see a Razorpay logo on admin save settings page. Where is the logo supposed to be?

File: admin/view/template/payment/razorpay.twig [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+			<div class="alert alert-danger">
+				<i class="fa fa-exclamation-circle"></i>
+				{{ error_warning }}
+				<button type="button" class="close" data-dismiss="alert">&times;</button>
+			</div>
+		{% endif %}
+		<div class="card">
+			<div class="card-header">
+				<i class="fa fa-pencil"></i>
+				{{ text_edit }}
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply admin/view/template/payment/razorpay.twig:34]: Updated

File: catalog/controller/payment/razorpay.php [hunk 3]
```diff
@@ -1,1095 +1,1329 @@
...
+			//echo();
+            if ($this->cart->hasSubscription() and 
+            	$this->config->get('payment_razorpay_subscription_status'))
+            {
+                $this->load->model('extension/razorpay/payment/razorpay');
+
+                //validate for non-subscription product and if recurring is product for more than 1
+                $this->validate_non_recurring_products();
+
+                if ($this->cart->hasRecurringProducts() > 1)
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply catalog/controller/payment/razorpay.php:55]: @sofiya02 As discussed in meeting earlier, we are including the subscriptions code as long it doesn't interfere with the checkout flow. We can spend time on this later when checkout is released for Opencart 4.

File: admin/controller/payment/razorpay.php [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+			$this->response->setOutput($this->load->view('extension/razorpay/payment/razorpay', $data));
+		}
+		catch(\Exception $e) {
+			echo(json_encode($e->getMessage()));
+			echo(json_encode($e->getTrace()));
+		}
+	}
+
+	protected function validate()
+    {
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply admin/controller/payment/razorpay.php:184]: @sofiya02 tested this fn. It works fine.

File: catalog/controller/payment/razorpay.php [hunk 5]
```diff
@@ -1,1095 +1,1329 @@
...
+        }
+    }
+
+    /**
+     * Handling order.paid event
+     * @param array $data Webook Data
+     */
+    protected function orderPaid(array $data)
+    {
+        $this->log->write('In OP logs');
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply catalog/controller/payment/razorpay.php]: Updated, please check.

File: catalog/controller/payment/razorpay.php [hunk 6]
```diff
@@ -1,1095 +1,1329 @@
...
+    const SUBSCRIPTION_PAUSED       = 'subscription.paused';
+    const SUBSCRIPTION_RESUMED      = 'subscription.resumed';
+    const SUBSCRIPTION_CANCELLED    = 'subscription.cancelled';
+    const SUBSCRIPTION_CHARGED      = 'subscription.charged';
+    const PA_WEBHOOK_WAIT_TIME      = 17;
+    const OP_WEBHOOK_WAIT_TIME      = 20;
+    const HTTP_CONFLICT_STATUS      = 409;
+
+    // Set RZP plugin version
+    private $version = '5.1.0'; //change this
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply catalog/controller/payment/razorpay.php]: Updated to 6.0.0. Is it fine?

File: catalog/controller/payment/razorpay.php [hunk 7]
```diff
@@ -1,1095 +1,1329 @@
...
+     * @param array $data Webook Data
+     */
+    protected function orderPaid(array $data)
+    {
+        $this->log->write('In OP logs');
+        $payment_created_time = $data['payload']['payment']['entity']['created_at'];
+
+        if(time() < ($payment_created_time + self::OP_WEBHOOK_WAIT_TIME))
+        {
+            header('Status: 409 Webhook conflicts due to early execution.', true, self::HTTP_CONFLICT_STATUS);
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply catalog/controller/payment/razorpay.php:441]: Have identified the issue. Will be updating it

File: catalog/controller/payment/razorpay.php [hunk 8]
```diff
@@ -1,1095 +1,1329 @@
...
+        return $this->load->view('extension/razorpay/payment/razorpay', $data);
+	}
+
+	private function get_order_creation_data($order_id)
+    {
+        $order = $this->model_checkout_order->getOrder($this->session->data['order_id']);
+
+        $data = [
+            'receipt' => $order_id,
+            'amount' => $this->currency->format($order['total'], $order['currency_code'], $order['currency_value'], false) * 100,
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply catalog/controller/payment/razorpay.php]: Fixed

File: admin/view/template/payment/razorpay.twig [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+				<br/>
+				<form id="form-payment" action="{{ save }}" method="post" data-oc-toggle="ajax" enctype="multipart/form-data">
+					<div class="form-group">
+						<div class="row mb-3">
+							<label class="col-sm-2 col-form-label" for="input-key-id">
+								<span data-toggle="tooltip" title="{{ help_key_id }}">{{ entry_key_id }}</span>
+							</label>
+							<div class="col-sm-10">
+								<input type="text" name="payment_razorpay_key_id" value="{{ razorpay_key_id }}" placeholder="{{ entry_key_id }}" id="input-key-id" class="form-control"/>
+								{% if error_key_id %}
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply admin/view/template/payment/razorpay.twig:52]: Added validations on client and server side for empty key and invalid key respectively

File: install.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "Razorpay",
+  "code": "razorpay",
+  "version": "1.0",
+  "author": "OpenCart Ltd",
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply install.json]: Updated, please check

File: admin/controller/payment/razorpay.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply admin/controller/payment/razorpay.php:184]: okay, I have moved this fn call inside index fn. Found the issue, will update it

File: admin/controller/payment/razorpay.php [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+			{
+				$data['error_warning'] = '';
+			}
+
+			if (isset($this->error['payment_razorpay_key_id']))
+			{
+				$data['error_key_id'] = $this->error['payment_razorpay_key_id'];
+			}
+			else
+			{
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply admin/controller/payment/razorpay.php:58]: Added check [here](https://github.com/razorpay/razorpay-opencart/pull/127/commits/da4b809be36cfc4f2d8eccc08bf2c0f75d7a67a8)

File: admin/view/template/payment/razorpay.twig (hunk 1 above)
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply admin/view/template/payment/razorpay.twig:2]: Trying to find how I can write the path correctly from View/Template file

File: admin/view/template/payment/razorpay.twig (hunk 9 above)
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply admin/view/template/payment/razorpay.twig:52]: [here](https://github.com/razorpay/razorpay-opencart/pull/127/commits/da4b809be36cfc4f2d8eccc08bf2c0f75d7a67a8)

File: admin/view/template/payment/razorpay.twig (hunk 1 above)
[the remark this answers is not in the document]
Comment (sofiya02) [reply admin/view/template/payment/razorpay.twig:2]: is this fixed

File: catalog/controller/payment/razorpay.php [hunk 12]
```diff
@@ -1,1095 +1,1136 @@
...
+            $this->log->write($e->getMessage());
+            $this->session->data['error'] = $e->getMessage();
+            echo "<div class='alert alert-danger alert-dismissible'> Something went wrong. Unable to create Razorpay Order Id.</div>";
+            exit;
+        }
+
+        try
+        {
+            $webhookUpdatedAt = ($this->config->get('payment_razorpay_webhook_updated_at') >= 0 ? 
+									$this->config->get('payment_razorpay_webhook_updated_at') : null);
```
Comment (sofiya02) [inline catalog/controller/payment/razorpay.php:148]: php indentation

**Review by sofiya02 — CHANGES_REQUESTED** [review]

File: catalog/controller/payment/razorpay.php (hunk 12 above)
Comment (nikhilsharma1100) [reply catalog/controller/payment/razorpay.php]: updated

File: README.md [hunk 13]
```diff
@@ -1,51 +1,51 @@
...
+    - order.paid
+    - subscription.charged
+    - subscription.paused
+    - subscription.resumed
+    - subscription.cancelled
+    
+
+### Development
+
+- The `master` branch holds the plugin for OpenCart 3
```
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply README.md:43]: master will be the branch with latest code

[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply README.md:43]: Have branched out master into opencartv3-5.1.1

File: admin/view/template/payment/razorpay.twig (hunk 1 above)
[the remark this answers is not in the document]
Comment (nikhilsharma1100) [reply admin/view/template/payment/razorpay.twig:2]: yes

**Review by sofiya02 — APPROVED** [review]
LGTM

**Pull request merged**